### DISCUSSION TITLE
Intercept shutdown

### DISCUSSION THREAD

**Discussion by agraven** [emilk/egui#813] [Q&A] [answered]
Hi. I'm writing an application where a graceful shutdown can take upwards of 15 seconds. Currently I do the shutdown with `App::on_exit`, however, this has the issue that the gui will be frozen the entire time the app is shutting down, making for a very poor user experience. I was wondering if there was a was of intercepting the signal to quit so I can keep the app responsive while it's shutting down.

**Comment by emilk** [comment, marked as answer]
Good question - there isn't a better way to do this at the moment, except by trying to get your users to quit your app from within the app instead (i.e. add a "File" menu with a "Quit" button, and then you do the shutdown).

I've opened an issue on it: https://github.com/emilk/egui/issues/814

**Reply by agraven** [reply]
The only user is myself right now since I haven't made the application public yet. When I do I think it should also serve as a pretty solid example of communication back and forth with an async runtime on a separate thread, I read that's something examples of were missing.

**Comment by follower** [comment]
@agraven If you want to followup on this, I've added a comment on the issue https://github.com/emilk/egui/issues/814#issuecomment-955579607 about the apparent need to handle [`WindowEvent::CloseRequested`](https://docs.rs/winit/0.25.0/winit/event/enum.WindowEvent.html#variant.CloseRequested) separately to get this behavior.
